// @parcel/core/lib/requests/PathRequest.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createPathRequest;
exports.ResolverRunner = void 0;

function _diagnostic() {
  const data = _interopRequireWildcard(require("@parcel/diagnostic"));

  _diagnostic = function () {
    return data;
  };

  return data;
}

function _logger() {
  const data = require("@parcel/logger");

  _logger = function () {
    return data;
  };

  return data;
}

function _utils() {
  const data = require("@parcel/utils");

  _utils = function () {
    return data;
  };

  return data;
}

function _nullthrows() {
  const data = _interopRequireDefault(require("nullthrows"));

  _nullthrows = function () {
    return data;
  };

  return data;
}

function _path() {
  const data = _interopRequireDefault(require("path"));

  _path = function () {
    return data;
  };

  return data;
}

function _url() {
  const data = _interopRequireDefault(require("url"));

  _url = function () {
    return data;
  };

  return data;
}

function _querystring() {
  const data = _interopRequireDefault(require("querystring"));

  _querystring = function () {
    return data;
  };

  return data;
}

function _ReporterRunner() {
  const data = require("../ReporterRunner");

  _ReporterRunner = function () {
    return data;
  };

  return data;
}

function _Dependency() {
  const data = _interopRequireDefault(require("../public/Dependency"));

  _Dependency = function () {
    return data;
  };

  return data;
}

function _PluginOptions() {
  const data = _interopRequireDefault(require("../public/PluginOptions"));

  _PluginOptions = function () {
    return data;
  };

  return data;
}

function _ParcelConfig() {
  const data = _interopRequireDefault(require("../ParcelConfig"));

  _ParcelConfig = function () {
    return data;
  };

  return data;
}

function _ParcelConfigRequest() {
  const data = _interopRequireDefault(require("./ParcelConfigRequest"));

  _ParcelConfigRequest = function () {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const type = 'path_request';
const QUERY_PARAMS_REGEX = /^([^\t\r\n\v\f?]*)(\?.*)?/;

function createPathRequest(input) {
  return {
    id: input.id,
    type,
    run,
    input
  };
}

async function run({
  input,
  api,
  options
}) {
  let {
    config
  } = (0, _nullthrows().default)((await api.runRequest((0, _ParcelConfigRequest().default)())));
  let resolverRunner = new ResolverRunner({
    options,
    config: new (_ParcelConfig().default)(config, options.packageManager, options.inputFS, options.autoinstall)
  });
  let assetGroup = await resolverRunner.resolve(input);

  if (assetGroup != null) {
    api.invalidateOnFileDelete(assetGroup.filePath);
  }

  return assetGroup;
}

class ResolverRunner {
  constructor({
    config,
    options
  }) {
    _defineProperty(this, "config", void 0);

    _defineProperty(this, "options", void 0);

    _defineProperty(this, "pluginOptions", void 0);

    this.config = config;
    this.options = options;
    this.pluginOptions = new (_PluginOptions().default)(this.options);
  }

  async getThrowableDiagnostic(dependency, message) {
    let diagnostic = {
      message,
      origin: '@parcel/core'
    };

    if (dependency.loc && dependency.sourcePath != null) {
      diagnostic.filePath = dependency.sourcePath;
      diagnostic.codeFrame = {
        code: await this.options.inputFS.readFile(dependency.sourcePath, 'utf8'),
        codeHighlights: dependency.loc ? [{
          start: dependency.loc.start,
          end: dependency.loc.end
        }] : []
      };
    }

    return new (_diagnostic().default)({
      diagnostic
    });
  }

  async resolve(dependency) {
    let dep = new (_Dependency().default)(dependency);
    (0, _ReporterRunner().report)({
      type: 'buildProgress',
      phase: 'resolving',
      dependency: dep
    });
    let resolvers = await this.config.getResolvers();
    let pipeline;
    let filePath;
    let query = {};
    let validPipelines = new Set(this.config.getNamedPipelines());

    if ( // Don't consider absolute paths. Absolute paths are only supported for entries,
    // and include e.g. `C:\` on Windows, conflicting with pipelines.
    !_path().default.isAbsolute(dependency.moduleSpecifier) && dependency.moduleSpecifier.includes(':')) {
      [pipeline, filePath] = dependency.moduleSpecifier.split(':');

      if (!validPipelines.has(pipeline)) {
        if (dep.isURL) {
          // This may be a url protocol or scheme rather than a pipeline, such as
          // `url('http://example.com/foo.png')`
          return null;
        } else {
          throw new Error(`Unknown pipeline ${pipeline}.`);
        }
      }
    } else {
      if (dependency.isURL && dependency.moduleSpecifier.startsWith('//')) {
        // A protocol-relative URL, e.g `url('//example.com/foo.png')`
        return null;
      }

      filePath = dependency.moduleSpecifier;
    }

    if (dependency.isURL) {
      let parsed = _url().default.parse(filePath);

      if (typeof parsed.pathname !== 'string') {
        throw new Error(`Received URL without a pathname ${filePath}.`);
      }

      filePath = decodeURIComponent(parsed.pathname);
    }

    let matchesQuerystring = filePath.match(QUERY_PARAMS_REGEX);

    if (matchesQuerystring && matchesQuerystring.length > 2) {
      filePath = matchesQuerystring[1];
      query = matchesQuerystring[2] ? _querystring().default.parse(matchesQuerystring[2].substr(1)) : {};
    }

    let diagnostics = [];

    for (let resolver of resolvers) {
      try {
        let result = await resolver.plugin.resolve({
          filePath,
          dependency: dep,
          options: this.pluginOptions,
          logger: new (_logger().PluginLogger)({
            origin: resolver.name
          })
        });

        if (result) {
          if (result.isExcluded) {
            return null;
          }

          if (result.filePath != null) {
            var _pipeline;

            return {
              canDefer: result.canDefer,
              filePath: result.filePath,
              query,
              sideEffects: result.sideEffects,
              code: result.code,
              env: dependency.env,
              pipeline: (_pipeline = pipeline) !== null && _pipeline !== void 0 ? _pipeline : dependency.pipeline,
              isURL: dependency.isURL
            };
          }

          if (result.diagnostics) {
            if (Array.isArray(result.diagnostics)) {
              diagnostics.push(...result.diagnostics);
            } else {
              diagnostics.push(result.diagnostics);
            }
          }
        }
      } catch (e) {
        // Add error to error map, we'll append these to the standard error if we can't resolve the asset
        let errorDiagnostic = (0, _diagnostic().errorToDiagnostic)(e, resolver.name);

        if (Array.isArray(errorDiagnostic)) {
          diagnostics.push(...errorDiagnostic);
        } else {
          diagnostics.push(errorDiagnostic);
        }

        break;
      }
    }

    if (dep.isOptional) {
      return null;
    }

    let dir = dependency.sourcePath != null ? (0, _utils().escapeMarkdown)((0, _utils().relativePath)(this.options.projectRoot, dependency.sourcePath)) : '';
    let specifier = (0, _utils().escapeMarkdown)(dependency.moduleSpecifier || ''); // $FlowFixMe because of the err.code assignment

    let err = await this.getThrowableDiagnostic(dependency, `Failed to resolve '${specifier}' ${dir ? `from '${dir}'` : ''}`); // Merge diagnostics

    err.diagnostics.push(...diagnostics);
    err.code = 'MODULE_NOT_FOUND';
    throw err;
  }

}

exports.ResolverRunner = ResolverRunner;
